File: src/Java/myservicebus/src/main/java/com/myservicebus/InMemoryJobService.java
package com.myservicebus;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CancellationException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import com.myservicebus.di.ServiceProvider;
import com.myservicebus.di.ServiceScope;
import com.myservicebus.tasks.CancellationToken;
import com.myservicebus.tasks.CancellationRegistration;
import com.myservicebus.tasks.CancellationTokenSource;

final class InMemoryJobService implements JobProvider {
    private static final class Entry {
        final UUID jobId;
        final Object job;
        final JobConsumerRegistry.Descriptor descriptor;
        final Instant submittedAtUtc;
        final Instant scheduledForUtc;
        final UUID recurringJobOccurrenceId;
        final List<JobAttemptState> attempts = new ArrayList<>();
        JobStatus status;
        Instant startedAtUtc;
        Instant completedAtUtc;
        Instant updatedAtUtc;
        JobProgress progress;
        CancellationTokenSource cancellation = new CancellationTokenSource();
        UUID scheduleToken;

        Entry(UUID jobId, Object job, JobConsumerRegistry.Descriptor descriptor,
                Instant submittedAtUtc, Instant scheduledForUtc, UUID recurringJobOccurrenceId, JobStatus status) {
            this.jobId = jobId;
            this.job = job;
            this.descriptor = descriptor;
            this.submittedAtUtc = submittedAtUtc;
            this.scheduledForUtc = scheduledForUtc;
            this.recurringJobOccurrenceId = recurringJobOccurrenceId;
            this.status = status;
            this.updatedAtUtc = submittedAtUtc;
        }
    }

    private final Map<UUID, Entry> jobs = new ConcurrentHashMap<>();
    private final JobConsumerRegistry registry;
    private final ServiceProvider services;
    private final LocalDelayScheduler delayScheduler;

    InMemoryJobService(JobConsumerRegistry registry, ServiceProvider services, LocalDelayScheduler delayScheduler) {
        this.registry = registry;
        this.services = services;
        this.delayScheduler = delayScheduler;
    }

    @Override
    public String getProviderName() {
        return "in-memory";
    }

    @Override
    public SchedulingDurability getDurability() {
        return SchedulingDurability.VOLATILE;
    }

    @Override
    public SchedulingPlacement getPlacement() {
        return SchedulingPlacement.PROCESS_LOCAL;
    }

    @Override
    public boolean isAuthoritative() {
        return true;
    }

    @Override
    public <TJob> CompletionStage<JobSubmissionReceipt> submit(
            TJob job,
            JobSubmissionOptions options,
            CancellationToken cancellationToken) {
        requireJob(job);
        cancellationToken.throwIfCancelled();
        Instant now = Instant.now();
        Entry entry = createEntry(job, options, now, null, JobStatus.WAITING);
        CompletableFuture.runAsync(() -> execute(entry));
        return CompletableFuture.completedFuture(receipt(entry));
    }

    @Override
    public <TJob> CompletionStage<JobSubmissionReceipt> schedule(
            Instant startAtUtc,
            TJob job,
            JobSubmissionOptions options,
            CancellationToken cancellationToken) {
        requireJob(job);
        if (startAtUtc == null) {
            throw new IllegalArgumentException("startAtUtc must not be null");
        }
        cancellationToken.throwIfCancelled();
        Instant now = Instant.now();
        Entry entry = createEntry(job, options, now, startAtUtc, JobStatus.SCHEDULED);
        return delayScheduler.schedule(startAtUtc, ignored -> {
            synchronized (entry) {
                if (entry.status == JobStatus.CANCELLED) {
                    return CompletableFuture.completedFuture(null);
                }
                entry.status = JobStatus.WAITING;
                entry.updatedAtUtc = Instant.now();
            }
            return CompletableFuture.runAsync(() -> execute(entry));
        }, CancellationToken.none()).thenApply(token -> {
            entry.scheduleToken = token;
            return receipt(entry);
        });
    }

    @Override
    public CompletionStage<JobControlResult> cancel(UUID jobId, CancellationToken cancellationToken) {
        cancellationToken.throwIfCancelled();
        Entry entry = jobs.get(jobId);
        if (entry == null) {
            return CompletableFuture.completedFuture(new JobControlResult(JobControlOutcome.NOT_FOUND, null));
        }
        UUID scheduleToken;
        synchronized (entry) {
            if (isTerminal(entry.status)) {
                return CompletableFuture.completedFuture(
                        new JobControlResult(JobControlOutcome.UNCHANGED, entry.status));
            }
            entry.cancellation.cancel();
            scheduleToken = entry.scheduleToken;
            if (entry.status == JobStatus.SCHEDULED || entry.status == JobStatus.WAITING) {
                complete(entry, JobStatus.CANCELLED);
            }
        }
        CompletionStage<Boolean> cancellation = scheduleToken == null
                ? CompletableFuture.completedFuture(false)
                : delayScheduler.cancel(scheduleToken);
        return cancellation.thenApply(ignored -> new JobControlResult(JobControlOutcome.APPLIED, entry.status));
    }

    @Override
    public CompletionStage<JobControlResult> retry(UUID jobId, CancellationToken cancellationToken) {
        cancellationToken.throwIfCancelled();
        Entry entry = jobs.get(jobId);
        if (entry == null) {
            return CompletableFuture.completedFuture(new JobControlResult(JobControlOutcome.NOT_FOUND, null));
        }
        synchronized (entry) {
            if (entry.status != JobStatus.FAULTED && entry.status != JobStatus.CANCELLED) {
                return CompletableFuture.completedFuture(
                        new JobControlResult(JobControlOutcome.INVALID_STATE, entry.status));
            }
            entry.cancellation = new CancellationTokenSource();
            entry.status = JobStatus.WAITING;
            entry.completedAtUtc = null;
            entry.updatedAtUtc = Instant.now();
        }
        CompletableFuture.runAsync(() -> execute(entry));
        return CompletableFuture.completedFuture(new JobControlResult(JobControlOutcome.APPLIED, JobStatus.WAITING));
    }

    @Override
    public CompletionStage<List<JobState>> getSnapshot(int maximumCount, CancellationToken cancellationToken) {
        requireMaximum(maximumCount);
        cancellationToken.throwIfCancelled();
        return CompletableFuture.completedFuture(jobs.values().stream()
                .sorted(Comparator.comparing((Entry entry) -> entry.updatedAtUtc).reversed())
                .limit(maximumCount)
                .map(this::state)
                .toList());
    }

    @Override
    public CompletionStage<List<JobAttemptState>> getAttempts(
            UUID jobId,
            int maximumCount,
            CancellationToken cancellationToken) {
        requireMaximum(maximumCount);
        cancellationToken.throwIfCancelled();
        Entry entry = jobs.get(jobId);
        if (entry == null) {
            return CompletableFuture.completedFuture(List.of());
        }
        synchronized (entry) {
            int start = Math.max(0, entry.attempts.size() - maximumCount);
            return CompletableFuture.completedFuture(List.copyOf(entry.attempts.subList(start, entry.attempts.size())));
        }
    }

    private Entry createEntry(Object job, JobSubmissionOptions options, Instant submittedAtUtc,
            Instant scheduledForUtc, JobStatus status) {
        JobSubmissionOptions effectiveOptions = options == null ? new JobSubmissionOptions() : options;
        UUID jobId = effectiveOptions.jobId() == null ? UUID.randomUUID() : effectiveOptions.jobId();
        Entry entry = new Entry(
                jobId,
                job,
                registry.get(job.getClass()),
                submittedAtUtc,
                scheduledForUtc,
                effectiveOptions.recurringJobOccurrenceId(),
                status);
        if (jobs.putIfAbsent(jobId, entry) != null) {
            throw new IllegalStateException("Job '" + jobId + "' already exists");
        }
        return entry;
    }

    private void execute(Entry entry) {
        JobConsumerRegistry.Descriptor descriptor = entry.descriptor;
        for (int retryAttempt = 0;; retryAttempt++) {
            boolean acquired = false;
            try {
                descriptor.concurrency().acquire();
                acquired = true;
                if (entry.cancellation.isCancelled()) {
                    synchronized (entry) {
                        complete(entry, JobStatus.CANCELLED);
                    }
                    return;
                }

                Instant startedAtUtc = Instant.now();
                UUID attemptId = UUID.randomUUID();
                synchronized (entry) {
                    entry.status = JobStatus.RUNNING;
                    if (entry.startedAtUtc == null) {
                        entry.startedAtUtc = startedAtUtc;
                    }
                    entry.updatedAtUtc = startedAtUtc;
                    entry.attempts.add(new JobAttemptState(
                            attemptId, entry.jobId, retryAttempt, JobAttemptStatus.RUNNING,
                            startedAtUtc, null, null, null));
                }

                CancellationTokenSource attemptCancellation = new CancellationTokenSource();
                try (CancellationRegistration registration = entry.cancellation.token()
                        .onCancel(attemptCancellation::cancel);
                        ServiceScope scope = services.createScope()) {
                    JobExecutionContext context = new JobExecutionContext(
                            entry.jobId,
                            attemptId,
                            retryAttempt,
                            entry.job,
                            startedAtUtc,
                            attemptCancellation.token(),
                            progress -> {
                                synchronized (entry) {
                                    entry.progress = progress;
                                    entry.updatedAtUtc = Instant.now();
                                }
                            });
                    CompletionStage<Void> stage = descriptor.run(scope.getServiceProvider(), context);
                    scope.detach();
                    stage.toCompletableFuture().get(
                            descriptor.options().getJobTimeout().toMillis(), TimeUnit.MILLISECONDS);
                    synchronized (entry) {
                        finishAttempt(entry, attemptId, JobAttemptStatus.COMPLETED, null);
                        complete(entry, JobStatus.COMPLETED);
                    }
                    return;
                } catch (CancellationException exception) {
                    synchronized (entry) {
                        finishAttempt(entry, attemptId, JobAttemptStatus.CANCELLED, null);
                        complete(entry, JobStatus.CANCELLED);
                    }
                    return;
                } catch (Exception exception) {
                    Throwable failure = unwrap(exception);
                    if (entry.cancellation.isCancelled()) {
                        synchronized (entry) {
                            finishAttempt(entry, attemptId, JobAttemptStatus.CANCELLED, null);
                            complete(entry, JobStatus.CANCELLED);
                        }
                        return;
                    }
                    if (failure instanceof TimeoutException) {
                        attemptCancellation.cancel();
                    }
                    synchronized (entry) {
                        finishAttempt(entry, attemptId, JobAttemptStatus.FAULTED, failure);
                    }
                    if (retryAttempt >= descriptor.options().getRetryCount()) {
                        synchronized (entry) {
                            complete(entry, JobStatus.FAULTED);
                        }
                        return;
                    }
                    synchronized (entry) {
                        entry.status = JobStatus.WAITING;
                        entry.updatedAtUtc = Instant.now();
                    }
                    Duration delay = descriptor.options().getRetryDelay();
                    if (delay != null && !delay.isZero()) {
                        Thread.sleep(delay.toMillis());
                    }
                }
            } catch (InterruptedException exception) {
                Thread.currentThread().interrupt();
                synchronized (entry) {
                    complete(entry, JobStatus.CANCELLED);
                }
                return;
            } finally {
                if (acquired) {
                    descriptor.concurrency().release();
                }
            }
        }
    }

    private void finishAttempt(Entry entry, UUID attemptId, JobAttemptStatus status, Throwable failure) {
        for (int index = 0; index < entry.attempts.size(); index++) {
            JobAttemptState current = entry.attempts.get(index);
            if (current.attemptId().equals(attemptId)) {
                entry.attempts.set(index, new JobAttemptState(
                        current.attemptId(), current.jobId(), current.retryAttempt(), status,
                        current.startedAtUtc(), Instant.now(),
                        failure == null ? null : failure.getClass().getName(),
                        failure == null ? null : failure.getMessage()));
                return;
            }
        }
    }

    private void complete(Entry entry, JobStatus status) {
        Instant now = Instant.now();
        entry.status = status;
        entry.completedAtUtc = now;
        entry.updatedAtUtc = now;
    }

    private JobSubmissionReceipt receipt(Entry entry) {
        return new JobSubmissionReceipt(
                entry.jobId, entry.status, entry.submittedAtUtc, entry.scheduledForUtc);
    }

    private JobState state(Entry entry) {
        synchronized (entry) {
            return new JobState(
                    entry.jobId,
                    entry.descriptor.jobTypeName(),
                    entry.status,
                    getProviderName(),
                    getDurability(),
                    getPlacement(),
                    entry.submittedAtUtc,
                    entry.scheduledForUtc,
                    entry.startedAtUtc,
                    entry.completedAtUtc,
                    entry.progress,
                    entry.recurringJobOccurrenceId,
                    entry.updatedAtUtc);
        }
    }

    private static Throwable unwrap(Throwable failure) {
        if ((failure instanceof ExecutionException || failure instanceof CompletionException)
                && failure.getCause() != null) {
            return failure.getCause();
        }
        return failure;
    }

    private static void requireJob(Object job) {
        if (job == null) {
            throw new IllegalArgumentException("job must not be null");
        }
    }

    private static void requireMaximum(int maximumCount) {
        if (maximumCount <= 0) {
            throw new IllegalArgumentException("maximumCount must be greater than zero");
        }
    }

    private static boolean isTerminal(JobStatus status) {
        return status == JobStatus.COMPLETED || status == JobStatus.FAULTED || status == JobStatus.CANCELLED;
    }
}

File: src/MyServiceBus/InMemoryJobService.cs
using System.Collections.Concurrent;
using Microsoft.Extensions.DependencyInjection;

namespace MyServiceBus;

public sealed class JobExecutionContext
{
    private readonly Action<JobProgress> setProgress;
    private readonly DateTimeOffset startedAtUtc;

    public JobExecutionContext(
        Guid jobId,
        Guid attemptId,
        int retryAttempt,
        object job,
        CancellationToken cancellationToken,
        DateTimeOffset startedAtUtc,
        Action<JobProgress> setProgress)
    {
        JobId = jobId;
        AttemptId = attemptId;
        RetryAttempt = retryAttempt;
        Job = job;
        CancellationToken = cancellationToken;
        this.startedAtUtc = startedAtUtc;
        this.setProgress = setProgress;
    }

    public Guid JobId { get; }

    public Guid AttemptId { get; }

    public int RetryAttempt { get; }

    public object Job { get; }

    public CancellationToken CancellationToken { get; }

    public TimeSpan ElapsedTime => DateTimeOffset.UtcNow - startedAtUtc;

    public Task SetProgress(long value, long? limit, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        CancellationToken.ThrowIfCancellationRequested();
        setProgress(new JobProgress(value, limit));
        return Task.CompletedTask;
    }
}

internal sealed class InMemoryJobContext<TJob> : JobContext<TJob>
    where TJob : class
{
    private readonly JobExecutionContext context;

    public InMemoryJobContext(JobExecutionContext context, TJob job)
    {
        this.context = context;
        Job = job;
    }

    public Guid JobId => context.JobId;

    public Guid AttemptId => context.AttemptId;

    public int RetryAttempt => context.RetryAttempt;

    public TJob Job { get; }

    public TimeSpan ElapsedTime => context.ElapsedTime;

    public CancellationToken CancellationToken => context.CancellationToken;

    public Task SetProgress(long value, long? limit = null, CancellationToken cancellationToken = default) =>
        context.SetProgress(value, limit, cancellationToken);
}

internal sealed class InMemoryJobService : IJobProvider
{
    private sealed class Entry
    {
        public required Guid JobId { get; init; }
        public required object Job { get; init; }
        public required IRegisteredJobConsumer Descriptor { get; init; }
        public required DateTimeOffset SubmittedAtUtc { get; init; }
        public DateTimeOffset? ScheduledForUtc { get; init; }
        public JobStatus Status { get; set; }
        public DateTimeOffset? StartedAtUtc { get; set; }
        public DateTimeOffset? CompletedAtUtc { get; set; }
        public DateTimeOffset UpdatedAtUtc { get; set; }
        public JobProgress? Progress { get; set; }
        public CancellationTokenSource Cancellation { get; set; } = new();
        public Guid? ScheduleToken { get; set; }
        public List<JobAttemptState> Attempts { get; } = [];
        public object Sync { get; } = new();
    }

    private readonly ConcurrentDictionary<Guid, Entry> jobs = new();
    private readonly JobConsumerRegistry registry;
    private readonly IServiceScopeFactory scopeFactory;
    private readonly ILocalDelayScheduler delayScheduler;

    public InMemoryJobService(
        JobConsumerRegistry registry,
        IServiceScopeFactory scopeFactory,
        ILocalDelayScheduler delayScheduler)
    {
        this.registry = registry;
        this.scopeFactory = scopeFactory;
        this.delayScheduler = delayScheduler;
    }

    public string ProviderName => "in-memory";

    public SchedulingDurability Durability => SchedulingDurability.Volatile;

    public SchedulingPlacement Placement => SchedulingPlacement.ProcessLocal;

    string IJobSource.Provider => ProviderName;

    public bool Authoritative => true;

    public Task<JobSubmissionReceipt> Submit<TJob>(
        TJob job,
        JobSubmissionOptions? options = null,
        CancellationToken cancellationToken = default)
        where TJob : class
    {
        ArgumentNullException.ThrowIfNull(job);
        cancellationToken.ThrowIfCancellationRequested();
        var now = DateTimeOffset.UtcNow;
        var entry = CreateEntry(job, options, now, null, JobStatus.Waiting);
        _ = Execute(entry);
        return Task.FromResult(CreateReceipt(entry));
    }

    public async Task<JobSubmissionReceipt> Schedule<TJob>(
        DateTimeOffset startAtUtc,
        TJob job,
        JobSubmissionOptions? options = null,
        CancellationToken cancellationToken = default)
        where TJob : class
    {
        ArgumentNullException.ThrowIfNull(job);
        cancellationToken.ThrowIfCancellationRequested();
        var now = DateTimeOffset.UtcNow;
        var scheduled = startAtUtc.ToUniversalTime();
        var entry = CreateEntry(job, options, now, scheduled, JobStatus.Scheduled);
        entry.ScheduleToken = await delayScheduler.Schedule(
            scheduled.UtcDateTime,
            _ =>
            {
                lock (entry.Sync)
                {
                    if (entry.Status == JobStatus.Cancelled)
                        return Task.CompletedTask;
                    entry.Status = JobStatus.Waiting;
                    entry.UpdatedAtUtc = DateTimeOffset.UtcNow;
                }
                return Execute(entry);
            },
            CancellationToken.None).ConfigureAwait(false);
        return CreateReceipt(entry);
    }

    public async Task<JobControlResult> Cancel(Guid jobId, CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (!jobs.TryGetValue(jobId, out var entry))
            return new JobControlResult(JobControlOutcome.NotFound);

        Guid? scheduleToken;
        lock (entry.Sync)
        {
            if (entry.Status is JobStatus.Completed or JobStatus.Faulted or JobStatus.Cancelled)
                return new JobControlResult(JobControlOutcome.Unchanged, entry.Status);
            entry.Cancellation.Cancel();
            scheduleToken = entry.ScheduleToken;
            if (entry.Status is JobStatus.Scheduled or JobStatus.Waiting)
                Complete(entry, JobStatus.Cancelled);
        }

        if (scheduleToken is not null)
            await delayScheduler.Cancel(scheduleToken.Value).ConfigureAwait(false);
        return new JobControlResult(JobControlOutcome.Applied, entry.Status);
    }

    public Task<JobControlResult> Retry(Guid jobId, CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (!jobs.TryGetValue(jobId, out var entry))
            return Task.FromResult(new JobControlResult(JobControlOutcome.NotFound));

        lock (entry.Sync)
        {
            if (entry.Status is not (JobStatus.Faulted or JobStatus.Cancelled))
                return Task.FromResult(new JobControlResult(JobControlOutcome.InvalidState, entry.Status));
            entry.Cancellation.Dispose();
            entry.Cancellation = new CancellationTokenSource();
            entry.Status = JobStatus.Waiting;
            entry.CompletedAtUtc = null;
            entry.UpdatedAtUtc = DateTimeOffset.UtcNow;
        }
        _ = Execute(entry);
        return Task.FromResult(new JobControlResult(JobControlOutcome.Applied, JobStatus.Waiting));
    }

    public Task<IReadOnlyList<JobState>> GetSnapshotAsync(
        int maximumCount,
        CancellationToken cancellationToken = default)
    {
        if (maximumCount <= 0)
            throw new ArgumentOutOfRangeException(nameof(maximumCount));
        cancellationToken.ThrowIfCancellationRequested();
        IReadOnlyList<JobState> snapshot = jobs.Values
            .OrderByDescending(entry => entry.UpdatedAtUtc)
            .Take(maximumCount)
            .Select(CreateState)
            .ToArray();
        return Task.FromResult(snapshot);
    }

    public Task<IReadOnlyList<JobAttemptState>> GetAttemptsAsync(
        Guid jobId,
        int maximumCount,
        CancellationToken cancellationToken = default)
    {
        if (maximumCount <= 0)
            throw new ArgumentOutOfRangeException(nameof(maximumCount));
        cancellationToken.ThrowIfCancellationRequested();
        if (!jobs.TryGetValue(jobId, out var entry))
            return Task.FromResult<IReadOnlyList<JobAttemptState>>([]);
        lock (entry.Sync)
            return Task.FromResult<IReadOnlyList<JobAttemptState>>(entry.Attempts.TakeLast(maximumCount).ToArray());
    }

    private Entry CreateEntry<TJob>(
        TJob job,
        JobSubmissionOptions? options,
        DateTimeOffset submittedAtUtc,
        DateTimeOffset? scheduledForUtc,
        JobStatus status)
        where TJob : class
    {
        var jobId = options?.JobId ?? Guid.NewGuid();
        var entry = new Entry
        {
            JobId = jobId,
            Job = job,
            Descriptor = registry.Get(typeof(TJob)),
            SubmittedAtUtc = submittedAtUtc,
            ScheduledForUtc = scheduledForUtc,
            Status = status,
            UpdatedAtUtc = submittedAtUtc
        };
        if (!jobs.TryAdd(jobId, entry))
            throw new InvalidOperationException($"Job '{jobId}' already exists.");
        return entry;
    }

    private async Task Execute(Entry entry)
    {
        var descriptor = entry.Descriptor;
        for (var retryAttempt = 0; ; retryAttempt++)
        {
            try
            {
                await descriptor.Concurrency.WaitAsync(entry.Cancellation.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                lock (entry.Sync)
                    Complete(entry, JobStatus.Cancelled);
                return;
            }

            try
            {
                if (entry.Cancellation.IsCancellationRequested)
                {
                    lock (entry.Sync)
                        Complete(entry, JobStatus.Cancelled);
                    return;
                }

                var startedAtUtc = DateTimeOffset.UtcNow;
                var attemptId = Guid.NewGuid();
                lock (entry.Sync)
                {
                    entry.Status = JobStatus.Running;
                    entry.StartedAtUtc ??= startedAtUtc;
                    entry.UpdatedAtUtc = startedAtUtc;
                    entry.Attempts.Add(new JobAttemptState(
                        attemptId,
                        entry.JobId,
                        retryAttempt,
                        JobAttemptStatus.Running,
                        startedAtUtc,
                        null,
                        null,
                        null));
                }

                using var timeout = new CancellationTokenSource(descriptor.Options.JobTimeout);
                using var linked = CancellationTokenSource.CreateLinkedTokenSource(
                    entry.Cancellation.Token,
                    timeout.Token);
                try
                {
                    using var scope = scopeFactory.CreateScope();
                    var context = new JobExecutionContext(
                        entry.JobId,
                        attemptId,
                        retryAttempt,
                        entry.Job,
                        linked.Token,
                        startedAtUtc,
                        progress =>
                        {
                            lock (entry.Sync)
                            {
                                entry.Progress = progress;
                                entry.UpdatedAtUtc = DateTimeOffset.UtcNow;
                            }
                        });
                    await descriptor.Run(scope.ServiceProvider, context).WaitAsync(linked.Token).ConfigureAwait(false);
                    lock (entry.Sync)
                    {
                        FinishAttempt(entry, attemptId, JobAttemptStatus.Completed, null);
                        Complete(entry, JobStatus.Completed);
                    }
                    return;
                }
                catch (OperationCanceledException) when (entry.Cancellation.IsCancellationRequested)
                {
                    lock (entry.Sync)
                    {
                        FinishAttempt(entry, attemptId, JobAttemptStatus.Cancelled, null);
                        Complete(entry, JobStatus.Cancelled);
                    }
                    return;
                }
                catch (Exception exception)
                {
                    var failure = timeout.IsCancellationRequested
                        ? new TimeoutException($"Job '{entry.JobId}' exceeded its timeout of {descriptor.Options.JobTimeout}.", exception)
                        : exception;
                    lock (entry.Sync)
                        FinishAttempt(entry, attemptId, JobAttemptStatus.Faulted, failure);

                    if (retryAttempt >= descriptor.Options.RetryCount)
                    {
                        lock (entry.Sync)
                            Complete(entry, JobStatus.Faulted);
                        return;
                    }

                    lock (entry.Sync)
                    {
                        entry.Status = JobStatus.Waiting;
                        entry.UpdatedAtUtc = DateTimeOffset.UtcNow;
                    }
                    if (descriptor.Options.RetryDelay is { } delay)
                        await Task.Delay(delay, entry.Cancellation.Token).ConfigureAwait(false);
                }
            }
            catch (OperationCanceledException)
            {
                lock (entry.Sync)
                    Complete(entry, JobStatus.Cancelled);
                return;
            }
            finally
            {
                descriptor.Concurrency.Release();
            }
        }
    }

    private static void FinishAttempt(
        Entry entry,
        Guid attemptId,
        JobAttemptStatus status,
        Exception? exception)
    {
        var index = entry.Attempts.FindIndex(attempt => attempt.AttemptId == attemptId);
        var current = entry.Attempts[index];
        entry.Attempts[index] = current with
        {
            Status = status,
            CompletedAtUtc = DateTimeOffset.UtcNow,
            FaultType = exception?.GetType().FullName,
            FaultMessage = exception?.Message
        };
    }

    private static void Complete(Entry entry, JobStatus status)
    {
        var now = DateTimeOffset.UtcNow;
        entry.Status = status;
        entry.CompletedAtUtc = now;
        entry.UpdatedAtUtc = now;
    }

    private static JobSubmissionReceipt CreateReceipt(Entry entry) =>
        new(entry.JobId, entry.Status, entry.SubmittedAtUtc, entry.ScheduledForUtc);

    private JobState CreateState(Entry entry)
    {
        lock (entry.Sync)
        {
            return new JobState(
                entry.JobId,
                entry.Descriptor.JobTypeName,
                entry.Status,
                ProviderName,
                Durability,
                Placement,
                entry.SubmittedAtUtc,
                entry.ScheduledForUtc,
                entry.StartedAtUtc,
                entry.CompletedAtUtc,
                entry.Progress,
                null,
                entry.UpdatedAtUtc);
        }
    }
}
